src/levanter/layerwise_trainer.py

```diff
@@ -0,0 +1,740 @@
+    TypeVar,
+    Union,
+)
+
+import equinox as eqx
+import haliax as hax
+import jax
+import jax.numpy as jnp
+import jmp
+import numpy as np
+from draccus import field
+from haliax import Axis
+from haliax.partitioning import ResourceAxis, ResourceMapping, named_jit
+from haliax.quantization import (
+    Fp8Config,
+    fp8_linear_layers,
+)
+from haliax.types import IntScalar, Scalar
+from jax.experimental import multihost_utils
+from jax.sharding import Mesh
+from jaxtyping import PRNGKeyArray, PyTree
+from optax import GradientTransformation, OptState
+
+import levanter.checkpoint
+import levanter.logging
+import levanter.tracker
+import levanter.tracker.wandb
+from levanter import tracker
+from levanter.checkpoint import (
+    CheckpointerConfig,
+    discover_latest_checkpoint,
+    load_checkpoint,
+    load_checkpoint_or_initialize,
+)
+from levanter.config import JsonAtom
+from levanter.data import (
+    Dataset,
+    ReplicatedBatchLoader,
+    ShardableDataset,
+    ShardedBatchLoader,
+)
+from levanter.distributed import DistributedConfig, RayConfig
+from levanter.logging import capture_time
+from levanter.models.lm_model import LmExample
+from levanter.tracker import TrackerConfig
+from levanter.trainer_state import (
+    _ensure_int_is_array,
+    cast_params_by_trainability,
+    init_optimizer_for_trainables,
+    saveable_training_mask,
+    take_train_step,
+    trainables_only,
+)
+from levanter.types import (
+    ComputeLossFunction,
+    FilterSpec,
+    FilterTree,
+    ModuleComputeLoss,
+)
+from levanter.utils import cloud_utils, fsspec_utils
+from levanter.utils.tree_utils import inference_mode
+
+logger = pylogging.getLogger(__name__)
+
+M = TypeVar("M")  # Model
+X = TypeVar("X")  # Input
+S = TypeVar("S")
+Student = TypeVar("Student")
+Teacher = TypeVar("Teacher")
+
+DEFAULT_JAX_CONFIG = {
+    "jax_threefry_partitionable": True,
+    "jax_softmax_custom_jvp": True,
+}
+
+
+def _per_layer_loss(inputs, layer, key, mask):
+    loss, teacher_x = inputs
+    key, layer_k = jax.random.split(key, 2)
+    student, teacher = layer
+    student_y = student(teacher_x, mask=mask, key=layer_k)
+    teacher_y = teacher(teacher_x, mask=mask, key=layer_k)
+    loss = hax.mean((teacher_y - student_y) ** 2)
+    return (loss, teacher_y)
+
+
+def _layer_loss(student, teacher, batch, key, compute_axis_mapping):
+    with hax.axis_mapping(compute_axis_mapping):
+        teacher_x = teacher.embeddings.embed(batch.tokens)
+        student_layers = student.transformer.layers.stacked
+        teacher_layers = teacher.transformer.layers.stacked
+        initial_loss = hax.NamedArray(0.0, axes=())
+
+        block_axis = student.transformer.layers.Block
+
+        def _is_block(leaf):
+            if hax.is_named_array(leaf):
+                print(leaf.shape, leaf.axes)
+            return hax.is_named_array(leaf) and leaf.axes[0] == block_axis
+
+        _loss_fn = hax.filter_checkpoint(_per_layer_loss)
+
+        loss, _ = hax.fold(
```

Comment:
Yeah this took a bit of back and forth to figure out but seems sensible now. 

Initially I  used an unrolled loop, which "worked" on a tiny model but effectively stalled in compilation for the larger model sizes.
When I switched to scan, I ran into a few issues, most the normal JAX/TPU stuff, which shouldn't affect most regular models:

* The `is_scanned` magic for `hax.fold` was trying to scan over the `initial` values in addition to the layers (because it by default tries to glom any namedarray into the scan). Maybe it's worth exempting the first argument from that logic?
* It took a while to realize I had to enable gradient checkpointing. Everything "works" on a CPU test, but just OOMs the process. Trying it on a TPU yields XLA errors messages about trying to allocate >1TB of memory and asking for magic flags to actually tell you the large tensors. (And of course, when you provide those flags, it doesn't change anything...). I ended up needing to reduce the size of the model to convince XLA I was close enough to fitting for it to actually bother to do the allocation analysis and cough up the errors...

I'm also obviously special-casing to the `stacked` variant of the transformer, which feels a little gross, but... I suspect it's not worth generalizing the `hof.py` modules to support this.